configuration/configuration.schema.json

```
@@ -0,0 +1,161 @@
+{
+  "title": "Arduino CLI configuration schema",
+  "description": "Describe the parameters available for the Arduino CLI configuration file. This schema should be considered unstable at this moment, it is not used by the CLI to validate input configuration",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "properties": {
+    "board_manager": {
+      "description": "",
+      "properties": {
+        "additional_urls": {
+          "description": "the URLs to any additional Boards Manager package index files needed for your boards platforms.",
+          "type": "array",
+          "items": {
+            "type": "string",
+            "format": "uri"
+          }
+        }
+      },
+      "type": "object"
+    },
+    "build_cache": {
+      "description": "configuration options related to the compilation cache",
+      "properties": {
+        "compilations_before_purge": {
+          "description": "interval, in number of compilations, at which the cache is purged, defaults to `10`. When `0` the cache is never purged.",
+          "type": "integer",
+          "minimum": 0
+        },
+        "ttl": {
+          "description": "cache expiration time of build folders. If the cache is hit by a compilation the corresponding build files lifetime is renewed. The value format must be a valid input for time.ParseDuration(), defaults to `720h` (30 days)",
+          "oneOf": [
+            {
+              "type": "integer",
+              "minimum": 0
+            },
+            {
+              "type": "string",
+              "pattern": "^[-+]?([0-9]*(\\.[0-9]*)?[a-z]+)+$"
```

```suggestion
              "pattern": "^\\+?([0-9]?\\.?[0-9]+(([nuµm]?s)|m|h))+$"
```

- Disallow negative values (allowed by `time.ParseDuration`, but not valid in this application)
- Require mandatory integer component (e.g., `h` is not a valid value)
- Require valid units (e.g., `1H` is not a valid value)

---

I'm fine with the regex, but I'm reluctant to implement strict validation within the scope of the issue, for this and other keys of the file.
The validation is correct and blameless, but it fails for current configurations that are not fed into the schema: the executable is less severe.
I'd avoid going to much into semantics, keeping it on the structural level at this stage. If and when we will have a sole logic, it will be a better time to tighten the schema.
Wdyt?

---

>it fails for current configurations

Please provide an example.

---

Example`-1H` is accepted, or better said tolerated, as a ttl currently.
In general most of the config keys are not validated and are passed down until something eventually fails, like `metrics.enabled: "maybe"` or a negative daemon port, which is not a problem as long as the daemon mode isn't used.

---

>most of the config keys are not validated

That is the whole point of the schema. What possible reason is there for wanting the schema to consider a value of `-1H` valid?

I'm sorry, but your regular expression is objectively wrong and I'm not going to approve it. I'm not even asking you to write a correct regular expression. I wrote it for you.

---

All good, sorry for the caused frustration